spring-rabbit/src/main/java/org/springframework/amqp/rabbit/support/CorrelationData.java

```diff
@@ -28,18 +28,23 @@
  */
 public class CorrelationData {
 
-	private String id;
+	private volatile String id;
 
 	public CorrelationData(String id) {
 		this.id = id;
 	}
 
 	public String getId() {
-		return id;
+		return this.id;
+	}
+
+	protected void setId(String id) {
```

Comment:
Well, I'm not going to convince you on the matter, just share my mind. Of course we can live with your current solution.
Just some extra noise for review.

Since this class is `public`, no one is prohibited to extend it to make access to this setter, therefore no need to mock of the hide-and-seek and just have this as `public` setter and change the `if` in the `CorrelationMessagePostProcessor` to `!= null` because it true by the `convertSendAndReceive()` logic.

Just my five fingers and even if you don't change the code, I'll merge your solution as is.